### ISSUE TITLE
dsc_sqlconfiguration reports an error.

### ISSUE THREAD

**Issue by MaciejWagner** [puppetlabs/Puppet.Dsc#225]
## Describe the Bug
I am trying to set SQL Server options using sqlserverdsc module. Unfortunetely I get the error from dsc_base_provider.
**My manifest:**
>   $instance = 'MSSQLSERVER'
>   $setting  =  'xp_cmdshell'
> 
>   dsc_sqlconfiguration { "${instance}/${setting}":
>     dsc_instancename => $instance,
>     dsc_optionname   => $setting,
>     dsc_optionvalue  => 0,
>   }

**An error:**

> Error: /Stage[main]/Main/Dsc_sqlconfiguration[MSSQLSERVER/xp_cmdshell]: Could not evaluate: undefined method `to_a' for "MSSQLSERVER/xp_cmdshell":String
> Did you mean?  to_c
>                to_r
>                to_f
>                to_i
>                to_s

## Expected Behavior
Puppet resource created successfully. SQL Server option changed successfully using dsc.

## Steps to Reproduce
1. Create the virtual machine with Windows Server and MSSQL Server for ex. from this Vagrant box: gusztavvargadr/sql-server
2. Install Puppet from: puppet-agent-6.4.2-x64.msi
3. Install modules: 
 - puppetlabs-chocolatey (v6.0.1)
 - puppetlabs-powershell (v5.0.0)
 - puppetlabs-pwshlib (v0.10.1)
 - puppetlabs-registry (v4.0.1)
 - puppetlabs-sqlserver (v3.0.0)
 - puppetlabs-stdlib (v8.1.0)
 - dsc-sqlserverdsc (v15.2.0-0-0)
4. Apply puppet manifest from the description

## Environment
 - puppetlabs-chocolatey (v6.0.1)
 - puppetlabs-powershell (v5.0.0)
 - puppetlabs-pwshlib (v0.10.1)
 - puppetlabs-registry (v4.0.1)
 - puppetlabs-sqlserver (v3.0.0)
 - puppetlabs-stdlib (v8.1.0)
 - dsc-sqlserverdsc (v15.2.0-0-0)
 - Platform Windows Server 2022 Standard Evaluation  OS build: 20348.405
 - Microsoft SQL Server 2019 (RTM-CU14) (KB5007182) - 15.0.4188.2 (X64)   Nov  3 2021 19:19:51   Copyright (C) 2019 Microsoft Corporation  Developer Edition (64-bit) on Windows Server 2022 Standard Evaluation 10.0 <X64> (Build 20348: ) (Hypervisor) 

## Additional Context
When I have listed hashes and catch from the failing function I had following results:
hashes:
MSSQLSERVER/xp_cmdshell
cache:
{:dsc_instancename=>"MSSQLSERVER", :dsc_optionname=>"xp_cmdshell", :dsc_optionvalue=>0, :dsc_restartservice=>false, :dsc_restarttimeout=>120, :dsc_servername=>"WINDB1542-SECUR", :name=>"MSSQLSERVER/xp_cmdshell"}

**Comment by chelnak** [comment]
Hey @Hvid & @MaciejWagner,

Is this still an issue for you? It's an interesting issue because as of Ruby 1.8 the string type does not have a `to_a` method.

This can be replicated easily with irb:

```ruby
irb(main):002:0> 'test'.to_a
Traceback (most recent call last):
        5: from /usr/bin/irb:23:in `<main>'
        4: from /usr/bin/irb:23:in `load'
        3: from /Library/Ruby/Gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):2
        1: from (irb):2:in `rescue in irb_binding'
NoMethodError (undefined method `to_a' for "test":String)
Did you mean?  to_r
               to_f
               to_i
               to_s
               to_c
```

Also, I see references to `to_a` in the base provider is on [line 34](https://github.com/puppetlabs/ruby-pwsh/blob/504909e0a6c7d0503d29cb340a19e240228988df/lib/puppet/provider/dsc_base_provider/dsc_base_provider.rb#L34). 

This makes me wonder whether we have a slight edge case with caching. I'm going to spend some time trying to replicate the issue with Puppet. However, if you have any more information that might help, let me know!

**Comment by chelnak** [comment]
Quick update from me:

I've actually managed to replicate this but only with the agent install. It works as expected when using the puppet gem.

OS: Windows Server 2022
Agent: 6.2.4
Gem: 6.2.4

This is some debug output from an agent run.

`hashes` (command [4]) appears to be an array of strings but we expect it to be an array of hashes. This will always cause the `NoMethodError: undefined method `to_a' for..` error when it's unpacked..

```bash
[1] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)>  
[2] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> item
=> {:dsc_ensure=>"Present", :dsc_loglevel=>"WarningsInfo", :dsc_logpath=>"C:\\Windows\\Logs\\DISM\\dism.log", :dsc_name=>"OpenSSH.Server~~~~0.0.1.0", :dsc_source=>nil, :name=>"ssh"}
[3] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> item.to_a
=> [[:dsc_ensure, "Present"], [:dsc_loglevel, "WarningsInfo"], [:dsc_logpath, "C:\\Windows\\Logs\\DISM\\dism.log"], [:dsc_name, "OpenSSH.Server~~~~0.0.1.0"], [:dsc_source, nil], [:name, "ssh"]]
[4] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> hashes
=> ["ssh"]
[5] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> hashes.select { |hash| hash.to_a }
NoMethodError: undefined method `to_a' for "ssh":String
Did you mean?  to_c
               to_r
               to_f
               to_i
               to_s
from (pry):4:in `block (2 levels) in fetch_cached_hashes'
```

When comparing that to a run using the puppet gem (same version), we do not get an error because `hashes` (command [4] again) is an array of hashes as expected:

```bash
[1] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)>  
[2] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> item
=> {:dsc_ensure=>"Present", :dsc_loglevel=>"WarningsInfo", :dsc_logpath=>"C:\\Windows\\Logs\\DISM\\dism.log", :dsc_name=>"OpenSSH.Server~~~~0.0.1.0", :dsc_source=>nil, :name=>"ssh"}
[3] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> item.to_a
=> [[:dsc_ensure, "Present"], [:dsc_loglevel, "WarningsInfo"], [:dsc_logpath, "C:\\Windows\\Logs\\DISM\\dism.log"], [:dsc_name, "OpenSSH.Server~~~~0.0.1.0"], [:dsc_source, nil], [:name, "ssh"]]
[4] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> hashes
=> [{:name=>"ssh", :dsc_name=>"OpenSSH.Server~~~~0.0.1.0"}]
[5] pry(#<Puppet::Provider::DscWindowscapability::DscWindowscapability>)> hashes.select { |hash| hash.to_a }
=> [{:name=>"ssh", :dsc_name=>"OpenSSH.Server~~~~0.0.1.0"}]
```

I think next steps for me would be to find out why this happens and whether or not it's expected behaviour from the agent. 

I'll update as I get more info!

**Comment by chelnak** [comment]
@Hvid @MaciejWagner 

**tldr**; you need to update to at least 6.7.x of the agent

Longer version..

It looks like you are seeing this issue because of the agent version you are currently running.

Agent 6.2.4 of the vendors an older version of the resource-api gem. This is what Puppet.Dsc uses under the hood to make the magic happen. 

This particular version of resource-api contains a bug that prevents it from handling composite namevars properly. Our generated DSC types contain two namevars, name and dsc_name) and therefore falls foul to the aforementioned bug.

This was fixed in version 1.8.6 of the resource-api gem and was vendored with the agent as of 6.7x.

My recommendation would be that you use the latest version of the agent, or at least a 6.7x variant to be able to benefit from the bugfix.

Hope this helps!
